eip-0012.md

```diff
@@ -0,0 +1,277 @@
+
+### get_balance(token_id: String = 'ERG') -> Value
+
+Returns available balance of {token_id} owned by the wallet. A useful and minimal convenience functionality, that while could be served by iterating the return of `get_utxos` and summing, is useful enough and could be most likely quickly returned by wallets without needing to do this calculation.
+
+### get_used_addresses(paginate: Paginate = 'undefined') -> [Address]
+
+Returns all used (exist in confirmed transactions) addresses of the user's wallet.
+If {paginate} is not `undefined` then results are returned in a paginated manner in chronological order of use on the blockchain. If {paginate} is out of range, `PaginateError` is returned.
+
+### get_unused_addresses() -> [Address]
+
+Returns unused addresses available that have not been used yet. The wallet would need to pre-generate addresses here within its restrictions (e.g. discover gap on indices for HD wallets). These can be used for receive addresses.
+
+### sign_tx(tx: Tx) -> Promise\<SignedTx> | TxSignError
+
+A signed tx is returned if the wallet can provide proofs for all inputs (P2PK ones). If not it produces an error. This should also have some kind of user pop-up which if rejected would produce an error as well. This way signing/keys are kept entirely in the wallet.
+
+### sign_tx_input(tx: Tx, index: number) -> Promise\<SignedInput> | TxSignError
+
+Lower level tx signing functionality that signs a single input if the wallet is able to, and returns an error if not. This can be useful for constructing transactions consisting of inputs from multiple parties, and should be general enough for future transaction signing use-cases. Likewise with `sign_tx` this should ask for user consent and can be rejected producing an error.
+
+### sign_data(addr: Address, message: String) -> Promise\<String> | DataSignError
+
+Signs the generic data {data} encoded as a hex string using the private key associated with {addr} if possible. If the wallet does not own this address, an error is returned. This should ask for user consent and produce an error if rejected. The wallet should also implement a message signing protocol such as the proposed [EmIP-005](https://github.com/Emurgo/EmIPs/blob/5b00fce84f31eb763892186eb9c88739ec809333/specs/emip-005.md) in order to make this endpoint safer for use, and make it harder for the user to accidentally sign data such as transactions, blocks, etc. Returns the signed data as a hex-encoded string.
+
+TBD: Hash algorithm/format to use.
+
+### submit_tx(tx: SignedTx) -> TxId | TxSendError
+
+Uses the wallet to send the given tx and returns the TxId for the dApp to track when/if it is included as its own responsibility. This is technically not mandatory but any wallet that is using such a bridge should have this functionality. Wallets can additionally choose to rate-limit this call, and can return an error if they did not or could not send the transaction.
+
+### add_external_box(box_id: BoxId)
```

Comment:
output parameter is missed

Reply:
There was originally none but perhaps this is more useful to return whether or not the wallet is tracking it.

Reply:
You mean whether it was tracked before the call?

Reply:
Whether it's tracked after the call regardless of whether it was tracked before. If it's tracked before and you call it, it returns true. If it was untracked before and it's tracked now it returns true. If it was untracked before and the wallet can't or won't track it, it returns false.